Make failed on macOS: Undefined symbols for architecture x86_64

The build fails on macOS:

```
[100%] Linking CXX executable nyancat
Undefined symbols for architecture x86_64:
  "nyan::Number<long long>::apply_value(nyan::Value const&, nyan::nyan_op)", referenced from:
      vtable for nyan::Number<long long> in nyan_tool.cpp.o
  "nyan::Number<long long>::allowed_operations(nyan::Type const&) const", referenced from:
      vtable for nyan::Number<long long> in nyan_tool.cpp.o
  "nyan::Number<long long>::str() const", referenced from:
      vtable for nyan::Number<long long> in nyan_tool.cpp.o
  "nyan::Number<long long>::copy() const", referenced from:
      vtable for nyan::Number<long long> in nyan_tool.cpp.o
  "nyan::Number<long long>::hash() const", referenced from:
      vtable for nyan::Number<long long> in nyan_tool.cpp.o
  "nyan::Number<long long>::repr() const", referenced from:
      vtable for nyan::Number<long long> in nyan_tool.cpp.o
  "nyan::Number<long long>::equals(nyan::Value const&) const", referenced from:
      vtable for nyan::Number<long long> in nyan_tool.cpp.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [nyan/nyancat] Error 1
make[1]: *** [nyan/CMakeFiles/nyancat.dir/all] Error 2
make: *** [all] Error 2
```

--- a/nyan/nyan_tool.cpp
+++ b/nyan/nyan_tool.cpp
@@ -45,7 +45,7 @@ void test_parser(const std::string &base_path, const std::string &filename) {
 	auto value = *root->get("test.First").get<Int>("member");
 
 	std::cout << "after change: First.member == "
-	          << value
+	          << value.str()
 	          << std::endl;
 
 	if (value != 24) {

--- a/nyan/value/number.cpp
+++ b/nyan/value/number.cpp
@@ -64,18 +64,6 @@ Float::Number(const IDToken &token) {
 
 
 template <typename T>
-Number<T>::Number(T value)
-	:
-	value{value} {}
-
-
-template <typename T>
-ValueHolder Number<T>::copy() const {
-	return {std::make_shared<Number>(*this)};
-}
-
-
-template <typename T>
 void Number<T>::apply_value(const Value &value, nyan_op operation) {
 	const Number &change = dynamic_cast<const Number &>(value);
 
@@ -98,31 +86,6 @@ void Number<T>::apply_value(const Value &value, nyan_op operation) {
 	default:
 		throw Error{"unknown operation requested"};
 	}
-}
-
-
-template <typename T>
-std::string Number<T>::str() const {
-	return std::to_string(this->value);
-}
-
-
-template <typename T>
-std::string Number<T>::repr() const {
-	return this->str();
-}
-
-
-template <typename T>
-size_t Number<T>::hash() const {
-	return std::hash<T>{}(this->value);
-}
-
-
-template <typename T>
-bool Number<T>::equals(const Value &other) const {
-	auto &other_val = dynamic_cast<const Number &>(other);
-	return this->value == other_val.value;
 }
 
 
@@ -168,5 +131,24 @@ const BasicType &Float::get_type() const {
 
 	return type;
 }
+
+// explicit instantiation of member functions
+template
+void Number<typename Int::storage_type>::apply_value(
+	const Value &value, nyan_op operation);
+
+template
+void Number<typename Float::storage_type>::apply_value(
+	const Value &value, nyan_op operation);
+
+template
+const std::unordered_set<nyan_op> &
+Number<typename Int::storage_type>::allowed_operations(
+	const Type &with_type) const;
+
+template
+const std::unordered_set<nyan_op> &
+Number<typename Float::storage_type>::allowed_operations(
+	const Type &with_type) const;
 
 } // namespace nyan

--- a/nyan/value/number.h
+++ b/nyan/value/number.h
@@ -2,6 +2,8 @@
 #pragma once
 
 
+#include <functional>
+
 #include "value.h"
 
 
@@ -16,12 +18,25 @@ template <typename T>
 class Number : public Value {
 public:
 	Number(const IDToken &token);
-	Number(T value);
+	Number(T value)
+		:
+		value{value} {}
 
-	ValueHolder copy() const override;
-	std::string str() const override;
-	std::string repr() const override;
-	size_t hash() const override;
+	ValueHolder copy() const override {
+		return {std::make_shared<Number>(*this)};
+	}
+
+	std::string str() const override {
+		return std::to_string(this->value);
+	}
+
+	std::string repr() const override {
+		return this->str();
+	}
+
+	size_t hash() const override {
+		return std::hash<T>{}(this->value);
+	}
 
 	const std::unordered_set<nyan_op> &allowed_operations(const Type &with_type) const override;
 	const BasicType &get_type() const override;
@@ -30,9 +45,13 @@ public:
 		return this->value;
 	}
 
+	using storage_type = T;
 protected:
 	void apply_value(const Value &value, nyan_op operation) override;
-	bool equals(const Value &other) const override;
+	bool equals(const Value &other) const override {
+		auto &other_val = dynamic_cast<const Number &>(other);
+		return this->value == other_val.value;
+	}
 
 	/**
 	 * Actual numerical value.
@@ -52,4 +71,4 @@ using Int = Number<int64_t>;
  */
 using Float = Number<double>;
 
-} // namespace std
+} // namespace nyan

